### ISSUE TITLE
Add some sort of "pre-check" to WriteFiles to minimize overwrites

### ISSUE THREAD

**Issue by daveaglick** [statiqdev/Statiq#303]
When WriteFiles writes multiple documents to the same location you can get strange file lock errors. Is there a way to pre-check output names or otherwise produce a better error (or wait) on file output conflicts?

**Comment by LokiMidgard** [comment]
Is it that you want to write the files sequencly but the old file is still locked, or does the write actually happens parallel?

Happens this Problem when using multiple WriteFile Modules or if one WriteFile Module tries to write a File multiple times.

One solution would be some method that generates unique file names where you write the file. After all Files are written (before wyam exits) all files would be renamed/moved to there actual name/location. Of course this has some drawback, not only it is more complex and so also more error prone. But also it could confuse the user that uses "normal" System.IO.\* Apis to check if a File exists.

**Comment by daveaglick** [comment]
I was actually thinking something even simpler for now. It's not so much fixing a bug as providing better error feedback. It stands to reason that if more than one document is being given to `WriteFiles` with the same output file, something went wrong somewhere along the line. Because WriteFiles overwrites any existing content, you'd never actually want to send documents with the same output file through the module.

I see two enhancements that could help with this:
- Keep track of the output file for each document as they get processed in parallel. Use a `ConcurrentDictionary` to store the output file as a key and the document (or maybe just the document source) as the value. If a document goes to add it's output to the dictionary, but that key already exists, display a warning or error to the user and report the two conflicting input documents.
- If the above happens it might be worthwhile to just stop the generation since the output will be undefined (in parallel operations, it's unclear which one would win). However, if we did want to allow it to proceed, we could also store a lock for the first write and then block on the second until the first one is done (and so on for the third conflict, etc.)

As I'm writing this, another thought came into my head. We could allow the `WriteFiles` module to support an append mode. If we do that, then we'll certainly want to support the ability to continue but block for pending writes to the same output file. I'll open a new issue for this possibility...
